lib/internal/modules/esm/hooks.js

```diff
@@ -528,7 +530,17 @@ class HooksProxy {
         debug('wait for signal from worker');
         AtomicsWait(this.#lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 0);
         const response = this.#worker.receiveMessageSync();
-        if (response.message.status === 'exit') { return; }
+        if (response.message.status === 'exit') {
+          // TODO: I do not understand why this is necessary.
+          // node \
+          //   --no-warnings --experimental-loader 'data:text/javascript,process.exit(42)'
+          //   ./test/fixtures/empty.js
+          // Does not trigger `this.#worker.on('exit', process.exit);`.
+          // I think it is because `makeSyncRequest` keeps waiting to see another
+          // message and blocks the thread from ANY other activity including the exit.
+          process.exit(response.message.body);
+          return;
+        }
```

Comment:
Why this change? It seems like it shouldn’t be required in order to achieve the loader chaining change which is the PR’s intent.

Also let’s please remove the commentary, we don’t commit stuff like that.

Reply:
If I recall when first creating this it came up in one of the hooks tests that would hang indefinitely. More code paths are shared now (loading goes through a much more "common" pipeline than it did previously) and there's a good chance that is the initial reason for this. However, I do believe this is code change above _is_ actually solving a real problem that cropped up.

It `makeSyncRequest` there is this code block:

```js
    debug('got sync response from worker', { method, args });
    if (response.message.status === 'never-settle') {
      process.exit(kUnfinishedTopLevelAwait);
    } else if (response.message.status === 'exit') {
      process.exit(response.message.body);
    }
```

And you can see that there _IS_ the same exit trap for `response.message.status === 'exit'`. However, to reach this code block we must first pass through `this.#waitForWorker()` which itself calls `this.#worker.receiveMessageSync();`. I believe the problem arises if the worker has exited inside of `this.#waitForWorker()` at which point calling `this.#worker.receiveMessageSync();` will just hang forever because the atomics trigger will never happen.

The `this.#worker.on('exit', process.exit);` is a bit of an illusion in this example because the thread is blocked on the atomics trigger, even though there is an `exit` event waiting to be delivered that delivery can never occur.

Reply:
☝️ @aduh95

Reply:
Originally the code only returning, right? It is indeed to give the `'exit'` event handler the opportunity to exit the main thread.

Reply:
@aduh95 is the change here okay then, or is it a problem?

Reply:
It's been reverted in https://github.com/nodejs/node/pull/48559/commits/2f13f67039ed20872fb665158bdcb8b971fe56a8 :)